daemon.zig

```diff
@@ -239,54 +239,84 @@ const DataSockHandler = struct {
     fn processSingle(self: *DataSockHandler, buf: []align(8) const u8) error{Handled}!usize {
         switch (self.state) {
             .auth => |*auth| {
-                const offsets = parseLineOffsets(buf);
+                var line_buf: []const u8 = buf;
+                var processed: usize = 0;
+
+                if (!self.first_byte_seen) {
+                    if (buf[0] == 0) {
+                        line_buf = buf[1..];
+                        processed += 1;
+                        self.first_byte_seen = true;
+                    } else {
+                        std.log.info("s={}: First byte of stream is not nul, disconnecting", .{self.sock});
+                        self.deinit();
+                        return error.Handled;
+                    }
+                }
+
```

Comment:
I'm having a hard time following and verifying these changes in my head.  We've split out some of the state into this new "first_byte_seen" field, but I'm not sure why it's needed since the previous version of the code was already verifying that the firstr byte was null by ensuring the data started with `\x00AUTH EXTERNAL`.  Splitting this exta `first_byte_seen` state into it's own field is telling me that each state can have that variable be set to `true` or `false` but that doesn't seem to be the case so I'm left a little unsure which states are valid and/or possible and what they each mean.  It might help me understand by maybe explaining the difference in behavior we are trying to implement here?  Was it possible before to send a message that didn't start with a null byte?

Reply:
The first message does can also be the `AUTH` command without parameters. This command queries the available authentication methods. It is then possible for the client to follow that up with an `AUTH EXTERNAL` command or `AUTH SKEY` etc. If we only match for `\x00AUTH EXTERNAL`, this follow-up command will not be detected as the null byte is only sent at the beginning of the stream. The null byte at the beginning is not part of the command and IMO shouldn't be treated as part of the command.

This is an example of this behavior when attempting to connect to the daemon (with this PR implemented) with d-feet:

```
info: bound socket to '/tmp/dbus-test'
info: s=5: new connection
info: s=5: got command 'AUTH'
info: s=5: got command 'AUTH EXTERNAL 31303030'
info: TODO: authenticate uid '31303030'
info: s=5: got command 'NEGOTIATE_UNIX_FD'
info: s=5: NEGOTIATE_UNIX_FD not implemented, sending ERROR
info: s=5: got command 'BEGIN'
```

I'm sure there is a better way of saving whether we are at the beginning of the stream and need to read that mandatory null byte (orelse disconnect), but IMO we shouldn't treat the null byte as part of the actual command.

Reply:
It's been a while since I implemented this and looked at the protocol so forgive me being slow to remember how this all works :)

If we always expect the first byte from a client to be null, I would expect something more like this: https://github.com/marler8997/dbus-zig/commit/27cca68f0b847c0ede19bb76527ad2229e289248

However, this change doesn't appear to work.  I don't think I've implemented the protocol correctly and it also causes an issue because it treats the first byte as it's own command which messes up the alignment on following messages.  I'll see if I can look into this more but if you see me missing something let me know.

Reply:
> It's been a while since I implemented this and looked at the protocol so forgive me being slow to remember how this all works :)

No problem :smile: 

> 
> If we always expect the first byte from a client to be null, I would expect something more like this: [27cca68](https://github.com/marler8997/dbus-zig/commit/27cca68f0b847c0ede19bb76527ad2229e289248)

You have a

```zig
self.state = .{ .auth = .{} };
```

missing in the switch prong for `.start` so that the state changes from start to auth.

> 
> However, this change doesn't appear to work. I don't think I've implemented the protocol correctly and it also causes an issue because it treats the first byte as it's own command which messes up the alignment on following messages. I'll see if I can look into this more but if you see me missing something let me know.

Yeah, regarding the alignment issue I think we cannot expect that `processSingle` will always only process a number of bytes that is a multiple of 8. In the authentication protocol, the messages are not guaranteed to be of a size divisible by 8 and while message headers have a size which is a multiple of 8, the message body doesn't have that restriction. Would it make sense for `process` to "clear" the buffer after `processSingle` returns (e.g. with a `memcpy`?)